workspaces/adventure-pack/goodies/java/src/traverse_post_order/AP.java

```diff
@@ -0,0 +1,63 @@
+package traverse_post_order;
+
+import iterable_stream.IterableStream;
+import java.util.NoSuchElementException;
+import java.util.Stack;
+import pair.Pair;
+import simple_iterator.SimpleIterator;
+
+public final class AP {
+
+  /**
+   * Get {@link IterableStream iterable stream} of all the {@link TreeNode nodes} in the binary tree
+   * in post-order.
+   * @param root Binary tree root {@link TreeNode node}.
+   * @return The {@link IterableStream iterable stream} in post-order.
+   */
+  public static IterableStream<TreeNode> traversePostOrder(
+    final TreeNode root
+  ) {
+    final Stack<Pair<TreeNode, Boolean>> stack = new Stack<
+      Pair<TreeNode, Boolean>
+    >();
+    stack.push(new Pair<>(root, false));
+    return IterableStream.from(
+      SimpleIterator.toIterator(() -> {
+        while (!stack.isEmpty()) {
+          Pair<TreeNode, Boolean> peek = stack.pop();
+          TreeNode node = peek.first();
+          boolean didTraverseChildren = peek.second();
+          if (node == null) continue;
+          if (didTraverseChildren) {
+            return node;
+          }
+          stack.push(new Pair<>(node, true));
```

Comment:
I think some strategic empty lines to separate different sections of the implementation can improve readability. For example the if statements handle the cases when we shouldn't push anything more to the stack, and the code here and below handles the case when we should.

What do you think? Once again I'd encourage consistency with the other Java traversal code, so if we come up with something better here, we should follow up and change it there as well.

Reply:
Yes. Strategic empty lines surely can help. I've made the necessary changes to make the spacing consistent with the code in java inorder-traversal goody.